src/App.js

```diff
@@ -22,50 +23,140 @@ const generateSquares = () => {
     }
   }
 
+  // console.log(squares);
+  // squares[1][1].value = PLAYER_1;
+
   return squares;
-}
+};
 
 const App = () => {
+  // Wave 2
+  // You will need to create a method to change the square
+  //   When it is clicked on.
+  //   Then pass it into the squares as a callback
 
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
+  console.log(squares);
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
 
-  // Wave 2
-  // You will need to create a method to change the square 
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
+  // console.log(squares);
+
+  // this is the function for the onClickCallback button-so when the user click the square, it calls updateSquareData
+  const updateSquareData = (updatedSquare) => {
+    console.log(squares);
+    // spread is copying the who array of squares
+    const newSquares = [...squares];
+    for (let row = 0; row < squares.length; row++) {
+      for (let col = 0; col < squares.length; col++) {
+        if (checkForWinner() === null) {
+          if (newSquares[row][col].id === updatedSquare.id) {
+            if (newSquares[row][col].value === '') {
+              newSquares[row][col].value = currentPlayer;
+              setCurrentPlayer(
+                currentPlayer === PLAYER_2 ? PLAYER_1 : PLAYER_2
+              );
+            }
+          }
+        }
+      }
+    }
+    setSquares(newSquares);
+    // if current player is player2 then it switch to player1
 
+    checkForWinner();
+  };
```

Comment:
You may want to check if `!checkForWinner()` is `falsy` before going through the for nested loops.  Additionaly, we can use the `!` (not) operator and combine the nested conditional using the `&&` (and) operator to reduce how nested this section is:


```suggestion
    if (!checkForWinner()) {
    for (let row = 0; row < squares.length; row++) {
      for (let col = 0; col < squares.length; col++) {
          if (newSquares[row][col].id === updatedSquare.id && newSquares[row][col].value === '') {
              newSquares[row][col].value = currentPlayer;
              setCurrentPlayer(
                currentPlayer === PLAYER_2 ? PLAYER_1 : PLAYER_2
              );
          }
        }
      }
    }
    setSquares(newSquares);
    checkForWinner();
  };
```